rules/no-typeof-undefined.js

```diff
@@ -0,0 +1,125 @@
+'use strict';
+const isShadowed = require('./utils/is-shadowed.js');
+const {matches} = require('./selectors/index.js');
+const {
+	addParenthesizesToReturnOrThrowExpression,
+} = require('./fix/index.js');
+const {removeSpacesAfter} = require('./fix/index.js');
+const isOnSameLine = require('./utils/is-on-same-line.js');
+const needsSemicolon = require('./utils/needs-semicolon.js');
+const {
+	isParenthesized,
+} = require('./utils/parentheses.js');
+
+const MESSAGE_ID_ERROR = 'no-typeof-undefined';
+const messages = {
+	[MESSAGE_ID_ERROR]: 'Compare with `undefined` directly instead of `typeof` check.',
+};
+
+const selector = [
+	'BinaryExpression',
+	matches(['===', '!==', '==', '!='].map(operator => `[operator="${operator}"]`)),
+	'[left.type="UnaryExpression"]',
+	'[left.operator="typeof"]',
+	'[left.prefix]',
+	'[right.type="Literal"]',
+].join('');
+
+/** @param {import('eslint').Rule.RuleContext} context */
+const create = context => {
+	const {
+		checkGlobalVariables,
+	} = {
+		checkGlobalVariables: true,
+		...context.options[0],
+	};
+
+	return {
+		[selector](binaryExpression) {
+			const {left: typeofNode, right: undefinedString} = binaryExpression;
+			if (undefinedString.value !== 'undefined') {
+				return;
+			}
+
+			const valueNode = typeofNode.argument;
+
+			if (
+				valueNode.type === 'Identifier'
+				&& !checkGlobalVariables
+				&& !isShadowed(context.getScope(), valueNode)
+			) {
+				return;
+			}
+
+			const sourceCode = context.getSourceCode();
+			const [typeofToken, secondToken] = sourceCode.getFirstTokens(typeofNode, 2);
+
+			return {
+				node: binaryExpression,
+				loc: typeofToken.loc,
+				messageId: MESSAGE_ID_ERROR,
+				* fix(fixer) {
+					// Change `==`/`!=` to `===`/`!==`
+					const {operator} = binaryExpression;
+					if (operator === '==' || operator === '!=') {
+						const operatorToken = sourceCode.getTokenAfter(
+							typeofNode,
+							token => token.type === 'Punctuator' && token.value === operator,
+						);
+
+						yield fixer.insertTextAfter(operatorToken, '=');
+					}
+
+					yield fixer.replaceText(undefinedString, 'undefined');
+
+					yield fixer.remove(typeofToken);
+					yield removeSpacesAfter(typeofToken, sourceCode, fixer);
+
+					const {parent} = binaryExpression;
+					if (
+						(parent.type === 'ReturnStatement' || parent.type === 'ThrowStatement')
+						&& parent.argument === binaryExpression
+						&& !isOnSameLine(typeofToken, secondToken)
+						&& !isParenthesized(binaryExpression, sourceCode)
+						&& !isParenthesized(typeofNode, sourceCode)
+					) {
+						yield * addParenthesizesToReturnOrThrowExpression(fixer, parent, sourceCode);
+						return;
+					}
+
+					const tokenBefore = sourceCode.getTokenBefore(binaryExpression);
+					if (needsSemicolon(tokenBefore, sourceCode, secondToken.value)) {
+						yield fixer.insertTextBefore(binaryExpression, ';');
+					}
```

Comment:
Same as https://github.com/sindresorhus/eslint-plugin-unicorn/blob/e4aaa42a780e2b155fa25fcba457162308dec36b/rules/no-negated-condition.js#L105, but I'm not going to make a reusable function in this PR, will do later.